src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,141 @@
+
+                if (matchingTypeDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, optionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Finds a matching type such that the display name of the type matches the name passed in, ignoring case. Case isn't used because
+            /// documents with name "Foo.cs" and "foo.cs" should still have the same type name
+            /// </summary>
+            private static async Task<SyntaxNode?> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf(n => !syntaxFacts.IsMethodBody(n)).Where(syntaxFacts.IsTypeDeclaration);
+                return typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(name, StringComparison.OrdinalIgnoreCase));
+            }
+
+            public static async Task<RenameSymbolDocumentAction?> TryCreateAsync(Document document, string newName, CancellationToken cancellationToken)
+            {
+                var analysis = await AnalyzeAsync(document, newName, cancellationToken).ConfigureAwait(false);
+
+                if (analysis.HasValue)
+                {
+                    return new RenameSymbolDocumentAction(analysis.Value);
+                }
+
+                return null;
+            }
+
+            private static async Task<AnalysisResult?> AnalyzeAsync(Document document, string newName, CancellationToken cancellationToken)
+            {
+                // TODO: Detect naming conflicts ahead of time
+                var originalSymbolName = Path.GetFileNameWithoutExtension(document.Name);
+                var matchingDeclaration = await GetMatchingTypeDeclarationAsync(document, originalSymbolName, cancellationToken).ConfigureAwait(false);
+
+                if (matchingDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingDeclaration, cancellationToken);
+                    var newSymbolName = Path.GetFileNameWithoutExtension(newName);
+
+                    if (symbol is null || symbol.Name == newSymbolName)
```

Comment:
why is this a case sensitive comparison when above it's insensitive?

Reply:
Fixed. Moved comparisons to utilities which makes it easier to follow